helix-view/src/gutter.rs

```diff
@@ -60,31 +62,42 @@ pub fn diagnostic<'doc>(
 
     Box::new(move |line: usize, _selected: bool, out: &mut String| {
         use helix_core::diagnostic::Severity;
-        if let Ok(index) = diagnostics.binary_search_by_key(&line, |d| d.line) {
-            let after = diagnostics[index..].iter().take_while(|d| d.line == line);
-
-            let before = diagnostics[..index]
-                .iter()
-                .rev()
-                .take_while(|d| d.line == line);
-
-            let diagnostics_on_line = after.chain(before);
-
-            // This unwrap is safe because the iterator cannot be empty as it contains at least the item found by the binary search.
+        if let Some(diagnostics_on_line) = diagnostic_by_line(diagnostics, line) {
             let diagnostic = diagnostics_on_line.max_by_key(|d| d.severity).unwrap();
 
             write!(out, "●").unwrap();
-            return Some(match diagnostic.severity {
+            Some(match diagnostic.severity {
                 Some(Severity::Error) => error,
                 Some(Severity::Warning) | None => warning,
                 Some(Severity::Info) => info,
                 Some(Severity::Hint) => hint,
-            });
+            })
+        } else {
+            None
         }
-        None
     })
 }
 
+pub fn diagnostic_by_line<'a>(
```

Comment:
Doesn't need to be pub. Also add doc comment

Reply:
Are you sure this needs a comment? Function name and return type looks pretty self-explanatory to me...